### DISCUSSION TITLE
How to acces the pipeline's ConfigLoader?

### DISCUSSION THREAD

**Discussion by mjmare** [kedro-org/kedro#1483] [Idea]
From the docs (https://kedro.readthedocs.io/en/stable/kedro_project_setup/configuration.html#local-and-base-configuration-environments) I know I can create a ConfigLoader. But how does one get the current ConfigLoader, ie the one set for the currently running pipeline?
For some more context: I want to access some credentials to initialise a hook. 

In Kedro 0.17 I did something like this:
```
from kedro.framework.session import get_current_session
       :
       :
        session = get_current_session()
        context = session.load_context()
        credentials = context.config_loader.get('*credentials.yml')
        rb_token = credentials['rollbar']['token']
        rb_environment = 'development' if context.env == 'local' else 'production'
        rollbar.init(rb_token,
                     environment=rb_environment,
                     )
        
```
But get_current_session has gone away in 0.18.0. So what is the Kedro-esque way of getting the current ConfigLoader?

**Comment by noklam** [comment]
In the coming `0.18.1` release, a [new hook](https://github.com/kedro-org/kedro/issues/1458) `after_context_created` will expose the `context` object and it contains the information like `env` and `config_loader`. 

https://github.com/kedro-org/kedro/issues/1458 There is a workaround in this issue for `0.18.0`.

**Reply by mjmare** [reply]
Thx. Could not discern the workaround though.

**Reply by noklam** [reply]
@mjmare Sorry, meant to link this issue instead. https://github.com/kedro-org/kedro/issues/1431.

`get_current_session()` was removed even before `0.18.0`, probably in `0.17.6` or `0.17.7`. The [workaround](https://github.com/Galileo-Galilei/kedro-mlflow/blob/c409432f6843c54bb739f6a5a67e8e871773086a/kedro_mlflow/config/kedro_mlflow_config.py#L233-L247) is using the private global variable `_active_session`. (This will be removed once the new hook is ready, which is likely in `0.18.1`, so please use it very cautiously). 

Please also be aware that you can't import the global variable directly due to how Python works, you need to import the module and call `module._active_session`, if you import the `_active_session` directly it become a local variable and it won't work as expected.

**Reply by mjmare** [reply]
Thanks. Workaround works!
Thanks also for the importing tip. It has bitten me in the past.

**Reply by giuliolovisotto** [reply]
Hi @noklam. What is the correct way to use the hook to access `context` from one of the nodes during a `kedro run` (I'm on `kedro==0.18.3`? I have something like the following:

```python
# My node function
def get_annotations(image_dataset: Dict[str, Callable]) -> List[Dict]:
    # what I am currently doing
    conf_source = str(Path.cwd() / settings.CONF_SOURCE)
    conf_loader = ConfigLoader(conf_source=conf_source, env="local")
    username = conf_loader.get("credentials.yml").get("username")
    # do something which includes username
    ...
```

Now that `get_current_session()` has been removed, is there a way to do this that does not require specifying the `conf_source` path with `Path.cwd()`?

Thanks for the help.

**Reply by noklam** [reply]
Accessing context from a node defies the purpose of having a context. This shouldn't be needed to access data. Why do you need to read credentials manually? Would a custom dataset be more appropriate?

**Reply by giuliolovisotto** [reply]
I must admit my setup may be strange and I'm new here (I found out about kedro recently and I really like it, I am trying to use it on a toy example). What I'd like to do in the node `get_annotations` is to use a remote API (Google Cloud) to do the processing of the `image_dataset`, then save the output. The init of the API requires credentials, including `username`, that's why I was trying to fetch these inside the node - would you do this differently? Thanks.

**Reply by noklam** [reply]
The IO Save and Load should be done by a DataSet and the processing logic will remains in the node.

https://kedro.readthedocs.io/en/stable/extend_kedro/custom_datasets.html

You can refer to other datasets implementation. If you are using GCS it would be supported by native Kedro datasets, otherwise you may need to implement the connection to the corresponding API.
